README.md
# PyPoster：轻量级的博客发布小工具
## 介绍
- **PyPoster** 是一个采用 Python 3.5 编写的博客离线发布小工具，GUI 采用 `tkinter` 框架构建。理论上，可以在安装了 Python 运行环境的多种平台下使用它。`PyPoster` 目前还比较简单，暂时只支持 WordPress 博客的 XML-RPC 接口。只需要指定本地的博客路径，即可一键发布！
- 项目地址：[https://github.com/chrisleegit/pyposter](https://github.com/chrisleegit/pyposter)。

## 为什么会有 PyPoster
以前在 Windows 平台下，会经常使用 WizNote 写东西，然后使用 WizNote 的博客发布功能进行发布。但是切换到 Ubuntu 或者 Mac OS 后，就只能使用 WizNote 的开源版本了，可惜那个版本没有 Windows 平台那么强大，所以也没有博客发布工具。所以就一直希望能有这么一款小工具，可以协助发布离线编写的文章。一开始是想找找有没有别人写的类似的小工具，但是并没有发现。由于平时主要使用 Ubuntu 系统，所以想找到一款可以在这个平台上运行又符合需求的小工具着实不易。于是 **PyPoster** 就诞生了。

## 已实现的功能
**Note**: 暂时只支持少量用到的功能，但是我觉得挺实用了。

- 支持命令行交互模式（测试使用）和 图形界面模式（强烈推荐使用带有图形界面的版本，操作简单）；
- 服务器信息自动保存，密码采用 `pycryto` 模块中提供的加密算法进行加密后再存储；
- 博客发布功能：新建博客或自动编辑博客（没有像 WizNote 那样可将 Markdown 文档选渲染后再发布的功能，主要是想用网站自定义的 Markdown 主题）；
- 博客发布时可以指定状态：__发布__ 和 __草稿__；
- 博客发布后，会自动添加 __版权信息__（该功能可移除）。

## 安装 & 运行
**Note：** 待功能稳定后，可能会发布打包好的版本，这样使用会更加方便。以下是在 Ubuntu 16.04 系统上安装并运行的步骤，其他系统类似。

- 请确保 [下载](https://github.com/ChrisLeeGit/pyposter/archive/master.zip) 最新的 PyPoster 项目代码；
- 确保系统安装好 Python 3.x 运行环境；
- 解压 PyPoster 源码包，并安装其中的 `requirements.txt` 中列出的依赖库：`pip3 install -r requirements.txt`；
- 进入 `src` 目录，运行 PyPoster：
    - `python3 pyposter.py`：进入 PyPoster 命令行交互模式；
    - `python3 pyposter_gui.py`：打开带有图形界面的 PyPoster。


## 编写并发布博客你的博客
1. 推荐为你要写的博客创建一个本地目录，假设叫做 `example_post`。
1. 使用你喜欢的文本编辑器编写 Markdown 格式的博客（PyPoster 就是设计用来处理 Markdown 的博客，当然也可以是其他类型的咯，如HTML文档等，反正它都会帮你发布）。
1. 如果你需要在博客中引用本地图片，你可以直接在博客文档中指定本地图片的相对路径或绝对路径（事实上，PyPoster 足够智能，它能够在成功上传图片后自动将博客中的图片路径替换成实际的 URL）。我推荐的做法是将图片都放在一个文件夹下（如与博客同目录的 `images` 子目录中），这样的结构会更加清晰。__需要注意的是，PyPoster 是按照图片文件的 HASH 值来确保图片的唯一性的，所以只要多张图片的 HASH 值相同，就会被当成同一张图片，从而确保不会重复上传图片。__
1. 当你完成博客后，目录结构可能和下面的类似：

    ```
    example_post/（存放博客文档和图片的目录）
    ├── images  （存放图片的目录）
    │   └── pyposter_gui.png    （图片文件）
    └── PyPoster, 轻量级的博客发布小工具.md    （博客文档）
    ```

1. GUI 版本发布博客：
    1. 启动图形化版本的 PyPoster； 
    1. 添加博客所在的路径（两种方式）：
        1. 方式一：在路径框中手动填写博客相对或者绝对路径；
        1. 方式二：点击左下角的 `添加` 按钮，找到博客文档并添加。
        
    1. 载入博客文档（如果使用方式二添加，则会博客文档会被自动载入）：点击 `载入` 按钮，程序会自动加载博客文档，如果存在配置文件，则会自动加载配置文件；
    1. 接下来，填写服务器信息：XML-RPC 地址，你的网站登录名和密码（第一次需要填写，以后会自动读取配置）；
    1. 选择发布状态：__草稿__ 和 __发布__；
    1. 给要发布的博客添加相应的标签（多个标签使用英文逗号隔开）和分类信息；你也可以点击按钮 `获取分类`，得到你的博客网站上所有分类，然后从中选择一个分类。
    1. 最后，待一切填写完毕，点击 `确定` 按钮，即可开始发布。可以在右侧的日志输出框看到详细的过程。博客发布完成后，会弹出提示框。
    1. 以下是运行截图（Ubuntu 16.04 系统）：
        - 加载相对路径博客：
        
        ![相对路径](tests/screenshots/relative_post_path.png)
        
        - 加载绝对路径博客（点击 `添加` 按钮选择更方便）：
        
        ![绝对路径](tests/screenshots/absolute_post_path.png)
        
        - 在 WordPress 网站上创建新的博客：
        
        ![创建](tests/screenshots/pyposter_new_post.png)
        
        - 编辑已经存在的博客，并将本地修改同步到 WordPress 网站：
        
        ![编辑](tests/screenshots/pyposter_edit_post.png)

1. 命令行版本发布博客（注意，最新的版本中已经不再支持命令行模式，相关测试代码也没有继续完善；由于使用命令行模式的频率较低，所以不再维护该功能）：
    1. `python3 pyposter.py` 启动 PyPoster；
    1. 根据提示填写信息，完成后即可发布。
    1. 命令行模式测试截图（旧版本测试截图）：
    
    ![Command](tests/screenshots/command_mode.png)

## 小小说明
- 第一次发布完成后，会在博客文档同目录下生成一个 名为`post.conf` 文件。那是一个非常重要的配置文件，与你的博客文档相关联（注意，不要将多个博客文档放在同一个目录，否则配置文件可能会相互覆盖）。每次在发布博客前，该配置文件都会被读取（如果没有会自动生成），然后会依据配置文件辨别哪些文件或内容发生了变动，从而防止图片或者博客重复发布。总之，一旦配置文件生成，就尽量不要删除，否则当你再次发布时，可能会再网站上看到相同的文章以及重复上传的图片。
- 每次博客或者图片发布成功后，都会在配置文件中记录对应的 HASH 值（目前采用 MD5 值），该 HASH 值是后期判断重复性操作的重要依据。
- 博客配置文件 `post.conf` 格式如下：

    ```
    {
        "posted_images": {
            "图片 HASH 值": [
                "图片本地路径",
                "图片发布后实际的URL"
            ]
        },
        "category": "博客分类",
        "tags": "标签1, 标签2",
        "checksum": "博客最后一次发布时的 HASH 值",
        "title": "博客标题（标题可以不同于博客实际的本地文件名）",
        "post_id": "博客发布后的 ID",
        "status": "publish/draft"
    }
    ```

## 依赖
- [python-wordpress-xmlrpc](https://github.com/maxcutler/python-wordpress-xmlrpc);
- [Python Cryptography Toolkit (pycrypto)](https://github.com/dlitz/pycrypto). 

## 改进
**如果你不小心看到了这里，并且对这个小工具感兴趣，愿意帮助完善的话，那将是一件令人兴奋的事情！ :)**

- 如果觉得有什么好的想法或者改进建议，请在 [issues](https://github.com/ChrisLeeGit/pyposter/issues) 中提出；
- 可以根据需要扩展功能；
- 如果有不错的小工具，也不妨推荐给我，多谢啦！我喜欢自动化的小工具，可以解放双手，节约时间。

## 主要参考
1. [python-wordpress-xmlrpc 文档](http://python-wordpress-xmlrpc.readthedocs.io/en/latest/)；
2. [pycrypto 相关模块 API 文档](https://www.dlitz.net/software/pycrypto/api/current/)；
3. [tkinter 主窗口居中显示](http://stackoverflow.com/questions/3352918/how-to-center-a-window-on-the-screen-in-tkinter)。

## 许可
[PyPoster](https://github.com/ChrisLeeGit/pyposter) 采用 [MIT](LICENSE.md) 许可协议。 


CHANGE_LOG.md
## 更新日志
-------------------------------
### 2016-07-21
- 新增 图片不存在时在日志输出框中显示，报告详细的路径；
- 修改 分级博客分类采用新的 树状控件 代替原先的 列表框控件，分类级别更清晰；

### 2016-07-20
- 新增 可以指定发布状态；
- 修复 没有发布情况下依然弹出打开链接提示框问题；
- 改善了部分代码，并添加了新的配置项。

### 2016-07-19 
#### v1.2
- 新增 自动添加版权信息（可取消该功能）。

#### v1.1
- 修复 分类列表框失焦后可能导致分类显示为空的问题；
- 新增 添加一个小尾巴（可以取消的）。

### 2016-07-18 v1.0
- 修复 GUI 版本加载配置文件时出现的问题；
- `pyposter` 转变成加载博客文档路径而非目录；
- 添加 MD5 文件校验，从而能够更准确地判断文件的变动情况；
- 修复 不同路径但可能为同一图片无法正确替换的问题；
- 修复 重复创建标签问题（英文标签不分大小写，不能重复创建）；
- 重构了部分函数，修改部分变量名称，避免歧义。

### 2016-07-17 v0.2
- 完成图形界面版本的布局，采用功能区和日志显示区的方式布局；
- 实现图形界面版本的 PyPoster 基本功能；
- 修复若干 bug；
- 更新样例。

### 2016-07-16 v0.1
- 实现 PyPoster 基本功能，可以正常新建和编辑博客；
- 图片上传功能实现；
- 命令行交互模式测试基本正常。


src/utils.py
#!/usr/bin/env python3
# -*-coding: utf-8-*-
# Author : Chris
# Blog   : http://blog.chriscabin.com
# GitHub : https://www.github.com/chrisleegit
# File   : utils.py
# Date   : 16-7-16
# Version: 0.1
# Description: utilities here.

import logging
from hashlib import md5
import os


def config_logger(log_file='log.txt', level=logging.INFO):
    """
    日志模块配置
    """
    file_fmt = '[%(asctime)s][%(module)-s][line %(lineno)-d][%(levelname)s] %(message)s'
    datefmt = '%m-%d %H:%M:%S'

    logging.basicConfig(
        level=level,
        format=file_fmt,
        datefmt=datefmt,
        filename=log_file,
        filemode='a')

    console = logging.StreamHandler()
    console.setLevel(level)

    console_fmt = '%(message)s'
    formatter = logging.Formatter(fmt=console_fmt,
                                  datefmt=datefmt)
    console.setFormatter(formatter)
    logging.getLogger('').addHandler(console)


def get_checksum(filename, blocksize=65536):
    if os.path.exists(filename):
        hasher = md5()
        with open(filename, 'rb') as f:
            buff = f.read(blocksize)
            while len(buff) > 0:
                hasher.update(buff)

                # get next block
                buff = f.read(blocksize)

        return hasher.hexdigest()


def get_text_checksum(text):
    hasher = md5()
    hasher.update(text.encode('utf-8'))
    return hasher.hexdigest()


def main():
    config_logger()
    print(get_text_checksum(open('../README.md').read()))


if __name__ == '__main__':
    main()


src/pyposter_gui.py
#!/usr/bin/env python3
# -*-coding: utf-8-*-
# Author : Chris
# Blog   : http://blog.chriscabin.com
# GitHub : https://www.github.com/chrisleegit
# File   : pyposter_gui.py
# Date   : 16-7-16
# Version: 0.1
# Description: pyposter_gui.

import json
import os
import sys
import webbrowser
from tkinter import Frame, LabelFrame, OptionMenu, Listbox, Scrollbar, Button, Entry, Label, Tk, StringVar, Radiobutton, \
    TclError
from tkinter.constants import *
from tkinter.scrolledtext import ScrolledText
from tkinter.ttk import Treeview, Style
from tkinter.messagebox import showinfo, showerror, askyesno
from tkinter.filedialog import askopenfilename
from utils import config_logger
from pyposter import PyPoster, PYPOSTER_PATH, LOG_PATH, ServerConfig, load_server_config, save_server_config
import logging
import threading

sys.path.append(PYPOSTER_PATH)

# 字体配置
FONT_DEFAULT = ('', 12, 'normal')
FONT_LOG = ('', 12, 'normal')


class OutputFrame(LabelFrame):
    def __init__(self, master=None, text="日志输出", cnf={}, **kw):
        super().__init__(master, cnf, **kw)
        self._output = ScrolledText(self, font=FONT_LOG, fg='blue')
        self._output.pack(fill=BOTH, expand=YES, padx=5)
        self.config(text=text, font=FONT_DEFAULT, padx=5)
        self.pack(side=RIGHT, fill=Y, expand=YES, padx=5, pady=5)

    def write(self, text):
        # self._output.config(state=NORMAL)
        self._output.insert(END, '{}'.format(text))
        self._output.see(END)
        # self._output.config(state=DISABLED)

    def read(self):
        pass

    def flush(self):
        pass

    def _get_line_count(self):
        return int(self._output.index(END).split('.')[0]) - 1

    def _clear(self):
        self._output.delete('1.0', 'end')


class PyPosterGUI(Frame):
    def __init__(self, master=None, cnf={}, **kwargs):
        super().__init__(master, cnf, **kwargs)
        self._function_frame = None
        self._output_frame = None
        self._tags = StringVar(self)
        self._post_title = StringVar(self)
        self._rpc_addr = StringVar(self)
        self._port = StringVar(self)
        self._username = StringVar(self)
        self._password = StringVar(self)
        self._post_path = StringVar(self)
        self._publish_status = StringVar(self)
        self._category_name = StringVar(self)
        self._pyposter = None
        self._categories = None
        self._make_widgets()
        self._load_config()
        self.pack(fill=Y, expand=YES, padx=5, pady=5, side=LEFT)
        self._center_window()
        self.master.protocol('WM_DELETE_WINDOW', self._on_closing)

    def _make_widgets(self):
        # 左边为功能区，右边为日志显示区
        # 分开布局
        self._function_frame = Frame(self)
        self._function_frame.pack(side=LEFT, fill=Y)

        self._make_post_info_frame()
        self._make_server_frame()
        self._make_publish_status_frame()
        self._make_options_frame()

        # 三个按钮
        btn_frm = Frame(self._function_frame)
        btn_frm.pack(side=BOTTOM, fill=X)

        # TO-DO: 添加按钮
        buttons = [
            ('添加', self._add_post_path, LEFT),
            ('载入', self._load_post, LEFT),
            ('确认', self._confirm, RIGHT)
        ]

        for col, btn in enumerate(buttons):
            Button(btn_frm, text=btn[0], font=FONT_DEFAULT, command=btn[1]).pack(fill=X, side=btn[2], padx=5, pady=5)

        # 日志显示
        self._output_frame = OutputFrame(self)

        # 重定向输出流
        sys.stderr = self._output_frame
        sys.stdout = self._output_frame

    def _make_options_frame(self):
        # 选项
        options_frm = LabelFrame(self._function_frame, text='选项', font=FONT_DEFAULT)
        options_frm.pack(fill=X, pady=5)
        # 博客标签
        Label(options_frm, text='标签 (逗号分隔)：', font=FONT_DEFAULT).grid(row=0, column=0, sticky=W, padx=5)
        Entry(options_frm, textvariable=self._tags, font=FONT_DEFAULT, width=30).grid(row=0, column=1, columnspan=2,
                                                                                      sticky=E, padx=5)
        # 博客分类
        Label(options_frm, text='博客分类：', font=FONT_DEFAULT).grid(row=1, column=0, sticky=W, padx=5)
        Entry(options_frm, textvariable=self._category_name, width=19, font=FONT_DEFAULT).grid(row=1, column=1, sticky=E)
        Button(options_frm, text='获取分类', font=FONT_DEFAULT, command=self._update_category).grid(row=1,
                                                                                                column=2,
                                                                                                sticky=E,
                                                                                                padx=5)
        cat_frm = Frame(options_frm)
        cat_frm.grid(row=2, column=0, columnspan=3, sticky=W, padx=5, pady=2)

        vs = Scrollbar(cat_frm)
        vs.pack(side=RIGHT, fill=Y)
        self._categories = Treeview(cat_frm, yscroll=vs.set)
        self._categories.bind('<ButtonRelease>', self._on_item_select)
        vs.config(command=self._categories.yview)
        self._categories.heading('#0', text='选择博客分类')
        self._categories.column('#0', width=400)
        s = Style()
        s.configure('.', font=FONT_DEFAULT)
        self._categories.pack(side=LEFT, fill=BOTH, expand=YES)

    def _make_operations_frame(self):
        # 发布操作
        # 自动，新建博客，编辑博客，新建页面，编辑页面
        pass

    def _make_publish_status_frame(self):
        # 发布状态
        status_frame = LabelFrame(self._function_frame, text='发布状态', font=FONT_DEFAULT)
        status_frame.pack(fill=X, pady=5)

        Radiobutton(status_frame, variable=self._publish_status, text='草稿',
                    value='draft', font=FONT_DEFAULT).pack(side=LEFT)

        Radiobutton(status_frame, variable=self._publish_status, text='发布',
                    value='publish', font=FONT_DEFAULT).pack(side=LEFT)

        self._publish_status.set('draft')

    def _make_server_frame(self):
        # 服务器信息
        server_frame = LabelFrame(self._function_frame, text='服务器信息', font=FONT_DEFAULT)
        server_frame.pack(fill=X, pady=5)
        # 一次创建完
        rows = [('RPC 地址：', self._rpc_addr, None),
                # ('端口：', self._port, None),
                ('用户名：', self._username, None),
                ('密码：', self._password, '*')]
        for index, row in enumerate(rows):
            Label(server_frame, text=row[0], font=FONT_DEFAULT).grid(row=index, column=0, sticky=W, padx=5, pady=2)
            Entry(server_frame, textvariable=row[1], show=row[2], font=FONT_DEFAULT, width=34).grid(row=index,
                                                                                                    column=1, sticky=E,
                                                                                                    padx=5, pady=2)

    def _make_post_info_frame(self):
        # 博客标题
        info_frm = LabelFrame(self._function_frame, text='博客信息', font=FONT_DEFAULT)
        info_frm.pack(side=TOP, fill=X, pady=5)

        rows = [
            ('博客路径：', self._post_path),
            ('博客标题：', self._post_title)
        ]

        for index, row in enumerate(rows):
            Label(info_frm, text=row[0], font=FONT_DEFAULT).grid(row=index, column=0, sticky=W,
                                                                 padx=5, pady=2)
            Entry(info_frm, textvariable=row[1], font=FONT_DEFAULT, width=35).grid(row=index, column=1, sticky=E,
                                                                                   padx=5, pady=2)

    def _center_window(self, width=850, height=610):
        screen_width = self.master.winfo_screenwidth()
        screen_height = self.master.winfo_screenheight()
        x = (screen_width / 2) - (width / 2)
        y = (screen_height / 2) - (height / 2)
        self.master.geometry('%dx%d+%d+%d' % (width, height, x, y))

    def _add_post_path(self):
        self._post_path.set(askopenfilename(title='添加博客路径...'))
        self._load_post()

    def _load_post(self):
        post_path = self._post_path.get()

        if not os.path.exists(post_path):
            logging.error('不存在的文件： {}'.format(post_path))
            showerror('错误！', '不存在的文件：{}'.format(post_path))
            return

        post_dir, post_filename = os.path.split(post_path)

        # 添加一个恰当的标题
        self._post_title.set(os.path.splitext(post_filename)[0])

        # 获取目录下的配置文件（如果没有，则无需加载）
        post_conf_path = os.path.join(post_dir, 'post.conf')
        if not os.path.exists(post_conf_path):
            logging.warning('博客配置文件未找到，采用默认配置！')
            return

        # 加载配置文件，读取相关信息，并填充到相应的控件中
        conf = json.load(open(post_conf_path))
        if conf:
            # 下面的做法可以防止已经填写好的信息被空的内容替代，那样是不好的。
            self._post_title.set(conf['title']) if conf['title'] != '' else None
            self._tags.set(conf['tags']) if conf['tags'] != '' else None
            self._category_name.set(conf['category']) if conf['category'] != '' else None
            self._publish_status.set(conf['status']) if conf['status'] != '' else None

    def _confirm(self):
        if not self._pyposter:
            self._init_pyposter()

        if self._pyposter:
            if not all([self._post_title.get(),
                        self._post_path.get()]):
                showerror('提示', '博客博客信息不完整，至少需要合法的博客路径！')
                return
            # 在后台线程中执行操作
            threading.Thread(target=self._post, daemon=True).start()

    def _post(self):
        link = self._pyposter.post(self._post_title.get(),
                                   self._category_name.get(),
                                   self._tags.get(),
                                   self._post_path.get(),
                                   self._publish_status.get())
        if link != '' and askyesno('提示', '博客发布成功，需要在浏览器中打开吗？'):
            webbrowser.open(link)

    def _update_category(self):
        if not self._pyposter:
            self._init_pyposter()

        if self._pyposter:
            # 删除旧的分类信息
            self._categories.delete(*self._categories.get_children())
            cates = self._pyposter.get_categories()
            logging.info('成功获取到分类：{} 个。'.format(len(cates)))
            while len(cates) > 0:
                # 每次成功添加一个节点，就把它从原来的表中移除，这样下次遍历时就会更快点，直到最后一个节点添加完毕。
                for cate in cates:
                    try:
                        if cate.parent == '0':
                            self._categories.insert('', 'end', cate.id, text=cate.name, open=True)
                        else:
                            self._categories.insert(cate.parent, 'end', cate.id, text=cate.name, open=True)
                        cates.remove(cate)
                    except TclError:
                        pass

    def _init_pyposter(self):
        if not self._is_server_info_valid():
            logging.error('服务器信息填写不完整！')
            showerror('错误！', '服务器信息填写不完整！')
        else:
            self._pyposter = PyPoster(self._rpc_addr.get(),
                                      self._username.get(),
                                      self._password.get())
            self._save_config()

    def _is_server_info_valid(self):
        return all([self._rpc_addr.get(), self._username.get(), self._password.get()])

    def _load_config(self):
        # 服务器信息
        server_config = load_server_config()
        if server_config:
            assert isinstance(server_config, ServerConfig)
            self._rpc_addr.set(server_config.rpc_address)
            self._username.set(server_config.username)
            self._password.set(server_config.password)

    def _save_config(self):
        save_server_config(ServerConfig(self._rpc_addr.get(),
                                        self._username.get(),
                                        self._password.get()))

    def _on_closing(self):
        if self._is_server_info_valid():
            self._save_config()
        self.quit()

    def _get_selected_category(self):
        try:
            focus_item = self._categories.focus()
            return self._categories.item(focus_item)['text']
        except: return ''

    def _on_item_select(self, event):
        # fix bug here
        cat = self._get_selected_category()
        if cat != '':
            self._category_name.set(cat)


def main():

    app = Tk()
    app.title('PyPoster, 博客发布小工具 (By Christopher L)。')
    app.resizable(width=False, height=False)

    # 左边是功能区
    PyPosterGUI(app)

    config_logger(LOG_PATH)
    logging.info('欢迎使用 PyPoster!')
    logging.info('详细的帮助文档请参见 README.md。')
    logging.info('Bug 反馈请访问项目主页：https://github.com/chrisleegit/pyposter。')
    logging.info('')
    app.mainloop()


if __name__ == '__main__':
    main()


src/pyposter.py
#!/usr/bin/env python3
# -*-coding: utf-8-*-
# Author : Chris
# Blog   : http://post.chriscabin.com
# GitHub : https://www.github.com/chrisleegit
# File   : pyposter.py
# Date   : 16-7-16
# Version: 0.1
# Description: PyPoster
import logging
import mimetypes
import os
import re
from wordpress_xmlrpc import *
from wordpress_xmlrpc.methods.media import UploadFile
from wordpress_xmlrpc.methods.posts import NewPost, EditPost, GetPost
from wordpress_xmlrpc.methods.users import GetUserInfo
from wordpress_xmlrpc.methods.taxonomies import *
from utils import config_logger, get_checksum, get_text_checksum
from json import load, dump
from pickle import load as p_load, dump as p_dump
import sys
from Crypto.Cipher import AES
from hashlib import md5


# 获取脚本所在目录
PYPOSTER_PATH = os.path.split(os.path.realpath(__file__))[0]

sys.path.append(PYPOSTER_PATH)

# PyPoster 服务器配置文件路径
SERVER_CONF_PATH = os.path.join(PYPOSTER_PATH, 'conf.pkl')

# 指定日志输出路径
LOG_PATH = os.path.join(PYPOSTER_PATH, 'log.txt')

# 添加小尾巴
LITTLE_TAIL = '\n\n[——来自 PyPoster，让博客发布更轻松！](https://github.com/ChrisLeeGit/pyposter)'

COPYRIGHT = "\n\n# 版权声明\n" \
            "1. 本文由 [Christopher L](http://blog.chriscabin.com) 发表，采用 [知识共享署名-非商业性使用-相同方式共享 4.0 国际许" \
            "可协议](http://creativecommons.org/licenses/by-nc-sa/4.0/) 进行许可。请确保你已了解许可协议，并在 __转载__ 时声明。\n" \
            "1. 本文固定链接：[{}]({})。\n{}"


class ServerConfig(object):
    def __init__(self, rpc_address, username, password):
        self._rpc_address = rpc_address
        self._username = username
        self._password = self._encrypt(password)

    @staticmethod
    def _encrypt(password):
        return AES.new(b'pyposter' * 2, AES.MODE_CFB, b'pyposter' * 2).encrypt(password)

    @staticmethod
    def _decrypt(password):
        return AES.new(b'pyposter' * 2, AES.MODE_CFB, b'pyposter' * 2).decrypt(password)

    @property
    def rpc_address(self):
        return self._rpc_address

    @property
    def username(self):
        return self._username

    @property
    def password(self):
        return (self._decrypt(self._password)).decode()

    def __str__(self):
        return 'rpc: {}\n' \
               'username: {}\n' \
               'password: {}\n'.format(self._rpc_address, self._username, self._password)


def load_server_config():
    # logging.info('Load pyposter config')
    if os.path.exists(SERVER_CONF_PATH):
        with open(SERVER_CONF_PATH, 'rb') as f:
            return p_load(f)


def save_server_config(config):
    # logging.info('Save pyposter config')
    if not config:
        return

    with open(SERVER_CONF_PATH, 'wb') as f:
        p_dump(config, f)


class PyPoster(object):
    def __init__(self, rpc_address, username, password):
        self._rpc_addr = rpc_address
        self._username = username
        self._password = password
        # 查找 markdown 中所有引用的图片
        self._image_re = re.compile(r'(!\[.*?\]\()(.+?)(\))')
        self._client = None
        self._post_conf = None
        self._login()

    def _login(self):
        try:
            self._client = Client(self._rpc_addr, self._username, self._password)
            print('登录成功！你好，{}！'.format(self._client.call(GetUserInfo())))
        except Exception as e:
            logging.error(str(e))
            sys.exit(1)

    def get_categories(self):
        if isinstance(self._client, Client):
            return self._client.call(GetTerms('category'))

    def get_post(self, post_id):
        try:
            return self._client.call(GetPost(post_id))
        except Exception as e:
            logging.error(str(e))
            return None

    def post(self, title, category, tags, post_path, status='publish', add_copyright=True):
        post_path = os.path.abspath(post_path)
        post_dir = os.path.split(post_path)[0]

        if not os.path.exists(post_path):
            logging.error('不存在的文件或目录：{}'.format(post_path))
            return

        # 切换工作目录，哈哈，这个是必须这样的
        old_cwd = os.getcwd()
        os.chdir(post_dir)

        # 读取博客内容
        content = open(post_path).read()
        if not content:
            return ''

        logging.info('准备发布博客：{}，请稍等...'.format(title))

        # 加载博客配置
        self._load_post_conf(post_dir)

        # 处理博客中所有引用的本地图片
        posted_images = self._process_images(content)

        # 处理博客文档内容
        content = self._process_post_content(content, posted_images)

        # 检查是否确实需要发布
        if not self._is_post_modified(title, category, tags, content, status):
            logging.warning('博客信息未更改，拒绝发布！再见～')
            return ''

        # 构建博客
        p = self._build_post(title, category, tags, content, status, add_copyright)

        if self._has_post():
            # 博客已经发布过了，此时只要编辑即可
            logging.info('开始编辑博客：{}({})'.format(title, self._post_conf['post_id']))
            p.id = self._post_conf['post_id']
            p.link = self._post_conf['link']
            self._client.call(EditPost(self._post_conf['post_id'], p))
        else:
            logging.info('开始新建博客：{}'.format(title))
            p.id = self._client.call(NewPost(p))
            self._post_conf['post_id'] = p.id

            # 拿到发布后的博文的 link
            fetched_post = self.get_post(p.id)
            p.link = fetched_post.link if fetched_post else ''
            self._post_conf['link'] = p.link

        if add_copyright:
            self._add_copyright(p)

        # 记录下文章的标题，分类和标签信息，可能会用到
        self._post_conf['category'] = category
        self._post_conf['tags'] = tags
        self._post_conf['title'] = title
        self._post_conf['checksum'] = get_text_checksum(content)
        self._post_conf['status'] = status

        # 最后要保存配置
        self._save_post_conf(post_dir)
        logging.info('博客发布完成！')

        # 切换回原来的工作目录
        os.chdir(old_cwd)

        return p.link

    @staticmethod
    def _process_post_content(content, posted_images):
        # 将内容中所有的图片地址替换成实际的url
        logging.info('处理博客内容...')
        for x in posted_images:
            content = content.replace(x[0], x[-1])

        return content

    def _process_images(self, content):
        # 提取文中引用到的图片
        images = self._get_all_valid_images(content)

        same_images = list()

        # 上传图片，成功上传图片后记录图片的md5值和实际URL
        for image in images:
            # 过滤重复图片，避免重复上传（使用md5值)
            checksum = get_checksum(image)
            if checksum not in self._post_conf['posted_images']:
                # 开始上传图片
                url = self._upload_image(image)
                if url:
                    # 添加到已经上传的列表
                    self._post_conf['posted_images'][checksum] = [image, url]
            else:
                # 检查是不是有可能存在不同路径但实际是同一张图片的情况
                posted_image = self._post_conf['posted_images'][checksum]
                if image != posted_image[0]:
                    logging.warning('发现相同的图片文件：{} 与 {} 的 HASH 值相同！将会自动替换为同一 URL！'.format(image, posted_image[0]))
                    same_images.append([image, posted_image[-1]])

        same_images.extend(list(self._post_conf['posted_images'].values()))
        return same_images

    def _build_post(self, title, category, tags, content, status, add_copyright):
        logging.info('构建博客：{}'.format(title))
        post = WordPressPost()
        post.title = title
        post.content = content + LITTLE_TAIL if not add_copyright else content

        # 发布状态
        post.post_status = status
        self._add_category(category, post)
        self._add_tags(post, tags)

        return post

    def _add_category(self, category, post):
        # 添加目录
        logging.info('添加目录：{}'.format(category))
        all_cats = self._client.call(GetTerms('category'))
        post_category = [x for x in all_cats if x.name == category]
        post.terms.append(post_category[0]) if len(post_category) == 1 else None

    def _add_tags(self, post, tags):
        # 添加标签，注意，标签是不分大小写的！！
        all_tags = self._client.call(GetTerms('post_tag'))
        tags_text = map(lambda x: x.strip(), tags.split(','))
        for tag_text in tags_text:
            # Bug fix: 空的 tag 不要添加
            if tag_text == '':
                continue

            logging.info('添加标签：{}'.format(tag_text))
            # 找 tag_text， 如果存在则无需创建，否则需要创建新的tag
            tag = [x for x in all_tags if x.name.lower() == tag_text.lower()]

            if len(tag) == 1:
                post.terms.append(tag[0])
            else:
                # 新建标签
                term = WordPressTerm()
                term.taxonomy = 'post_tag'
                term.name = tag_text
                term.id = self._client.call(NewTerm(term))
                post.terms.append(term)

    def _add_copyright(self, post):
        if not post:
            return
        logging.info('正在为博文添加版权信息，请耐心等待！')
        post.content = '{}{}'.format(post.content, COPYRIGHT.format(post.link, post.link, LITTLE_TAIL))
        self._client.call(EditPost(post.id, post))

    def _has_post(self):
        # 如果存在，则返回post_id
        return True if self._post_conf['post_id'] else False

    def _upload_image(self, imagename):
        # 准备数据
        data = dict()
        # 获取真实的文件名称，带有后缀名，不需要目录
        data['name'] = os.path.split(imagename)[-1]
        data['type'] = mimetypes.guess_type(imagename)[0]
        data['overwrite'] = 'false'
        with open(imagename, 'rb') as img:
            data['bits'] = xmlrpc_client.Binary(img.read())

        logging.info('上传图片：{}'.format(imagename))
        x = self._client.call(UploadFile(data))
        return x['url'] if x else None

    def _load_post_conf(self, post_dir):
        # 在博客路径下查找并加载配置文件，没有会自动生成
        conf_path = os.path.join(post_dir, 'post.conf')

        logging.info('加载博客配置文件{}'.format(conf_path))

        if os.path.exists(conf_path):
            self._post_conf = load(open(conf_path))
        else:
            self._post_conf = {'post_id': None, 'posted_images': dict(),
                               'category': '', 'tags': '', 'title': '',
                               'checksum': '', 'link': '', 'status': 'draft'}

    def _save_post_conf(self, post_dir):
        logging.info('保存博客配置到文件：{}'.format(os.path.join(post_dir, 'post.conf')))
        if self._post_conf:
            with open(os.path.join(post_dir, 'post.conf'), 'w') as f:
                dump(self._post_conf, f, indent=4)

    def _get_all_valid_images(self, content):
        assert isinstance(content, str)
        # 查找博客文档中所有引用的本地图片，会验证图片的有效性
        images = self._image_re.findall(content)
        result = list()
        for x in images:
            if os.path.exists(x[1]):
                result.append(x[1])
            else:
                logging.error('未找到图片：{}'.format(x[1]))

        return result

    def _is_post_modified(self, title, category, tags, content, status):
        # 如果状态变化应该发布
        if self._post_conf['status'] != status:
            return True

        # 检查博客是否发生变动了
        if title != self._post_conf['title']:
            return True

        if category != self._post_conf['category']:
            return True

        if tags != self._post_conf['tags']:
            return True

        if get_text_checksum(content) != self._post_conf['checksum']:
            return True

        return False


def main():
    config_logger()

    # 命令行模式
    from getpass import getpass
    print('>{} 欢迎使用 PyPoster！ {}<'.format('*' * 10, '*' * 10))

    wp = None

    # 如果存在配置文件，则无需输入
    conf = load_server_config()
    if conf:
        if not input('检查到服务器配置，是否需要重新配置？[Y(es)/N(o)]').lower() in ('y', 'yes'):
            wp = PyPoster(conf.rpc_address, conf.username, conf.password)

    if not wp:
        xml_rpc_address = input('xmlrpc 地址：')
        username = input('用户名：')
        password = getpass('密码：')
        wp = PyPoster(xml_rpc_address, username, password)

        # 保存配置
        save_server_config(ServerConfig(xml_rpc_address, username, password))

    while True:
        if 'quit' in input('输入"quit"退出，回车键继续：'):
            break
        else:
            post_path = input('博客路径：')
            post_title = input('博客标题：')
            category = input('博客分类：')
            tags = input('博客标签（用逗号隔开多个）：')
            if input('确认发布？ [Y(es)/N(o)]').lower() in ['y', 'yes']:
                wp.post(post_title, category, tags, post_path)


if __name__ == '__main__':
    main()